services/paymaster/src/api.rs

```diff
@@ -354,8 +357,58 @@ async fn sponsor_and_send_handler(
         confirmation_status.clone(),
     );
 
-    let gas = crate::constraint::compute_gas_spent(&transaction)?;
-    obs_gas_spend(domain, matched_variation_name, confirmation_status, gas);
+    // Spawn async task to fetch actual transaction costs from RPC
+    // This happens in the background to avoid blocking the response to the client
+    // Only fetch if the transaction was actually sent to chain (not rejected in preflight)
+    let should_fetch_costs = match &confirmation_result {
+        ConfirmationResult::Success { .. } => true,
+        ConfirmationResult::Failed { sent_to_chain, .. } => *sent_to_chain,
+    };
+
+    if should_fetch_costs {
+        let signature_str = match &confirmation_result {
+            ConfirmationResult::Success { signature } => signature,
+            ConfirmationResult::Failed { signature, .. } => signature,
+        };
+
+        if let Ok(signature) = signature_str.parse::<Signature>() {
```

Comment:
it seems strange that the signature is transformed from `Signature` to `String` and back to `Signature`.
Perhaps `ConfirmationResult` should use the type `Signature` and add the right serde configuration so that it becomes the right`String` when serialized?